source/Cosmos.HAL2/BlockDevice/AHCI.cs

```diff
@@ -0,0 +1,380 @@
+using System;
+using System.Collections.Generic;
+using Cosmos.HAL;
+using Cosmos.Debug.Kernel;
+using Cosmos.HAL.BlockDevice;
+using Cosmos.HAL.BlockDevice.Ports;
+using Cosmos.HAL.BlockDevice.Registers;
+using Cosmos.Core;
+using Cosmos.Core.Memory.Old;
+
+namespace Cosmos.HAL.BlockDevice
+{
+    public class AHCI
+    {
+        internal static Debugger mAHCIDebugger = new Debugger("HAL", "AHCI");
+        internal static PCIDevice xDevice = HAL.PCI.GetDevice(0x001, 0x0006);
+
+        private static List<StoragePort> mPorts = new List<StoragePort>();
+        private static GenericRegisters mGeneric;
+        private static ulong mABAR;
+
+        // Capabilities
+        #region Capabilities
+        private static bool Supports64bitAddressing;
+        private static bool SupportsNativeCommandQueuing;
+        private static bool SupportsSNotificationRegister;
+        private static bool SupportsMechanicalPresenceSwitch;
+        private static bool SupportsStaggeredSpinup;
+        private static bool SupportsAggressiveLinkPowerManagement;
+        private static bool SupportsActivityLED;
+        private static bool SupportsCommandListOverride;
+        private static uint InterfaceSpeedSupport;
+        private static bool SupportsAHCIModeOnly;
+        private static bool SupportsPortMutliplier;
+        private static bool FISBasedSwitchingSupported;
+        private static bool PIOMultipleDRQBlock;
+        private static bool SlumberStateCapable;
+        private static bool PartialStateCapable;
+        private static uint NumOfCommandSlots;
+        private static bool CommandCompletionCoalsecingSupported;
+        private static bool EnclosureManagementSupported;
+        private static bool SupportsExternalSATA;
+        private static uint NumOfPorts;
+        #endregion
+
+        // Constants
+        public const ulong RegularSectorSize = 512UL;
+
+        // Informations
+        public string SerialNo;
+        public string Version
+        {
+            get => ((byte)mGeneric.AHCIVersion >> 24) + (byte)(mGeneric.AHCIVersion >> 16) + "." + (byte)(mGeneric.AHCIVersion >> 8) + ((byte)(mGeneric.AHCIVersion) > 0 ? "." + (byte)mGeneric.AHCIVersion : "");
+        }
+
+        internal static void InitDriver()
+        {
+
+            if (xDevice != null)
+            {
+                AHCI Driver = new AHCI(xDevice);
+            }
+        }
+
+        internal PCIDevice GetDevice() => xDevice;
+
+        public AHCI(PCIDevice aAHCIDevice)
+        {
+            mAHCIDebugger.Send("Something wrong!");
```

Comment:
I think this can be removed nothing wrong if the Ctor is called :-)

Reply:
Tried my best to remove debug messages, but forgot xD